File: TopInterview150/Strings&Arrays/Arrays/GasStation/solution.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int total = 0;
        int curr = 0;
        int index = -1;
        for (int i = 0; i < gas.size(); i++)
        {
            curr += gas[i]-cost[i];
            total+= gas[i]-cost[i];
            if (curr >= 0 && index ==-1)
            {
                index = i;
            }
            if (curr < 0)
            {
                curr = 0;
                index = -1;
            }
        }
        if (total >= 0)
        {return index;}
        return -1;
    }
};

File: TopInterview150/Strings&Arrays/Arrays/GasStation/solution.py
class Solution(object):
    def canCompleteCircuit(self, gas, cost):
        """
        :type gas: List[int]
        :type cost: List[int]
        :rtype: int
        """
        totalGasMinusCost = 0
        currentGasMinusCost = 0
        index = 0
        active = False
        for x in range(len(gas)):
            totalGasMinusCost += gas[x] - cost[x]
            currentGasMinusCost += gas[x] - cost[x]
            if (currentGasMinusCost > 0 and not active):
                index = x
                active = True
            if (currentGasMinusCost < 0):
                index = -1
                active = False
                currentGasMinusCost = 0
        if totalGasMinusCost < 0:
            return -1
        return index
        